src/client/driver/command-executors/client-functions/selector-executor/node-snapshots.js

```diff
@@ -37,34 +37,44 @@ export class NodeSnapshot {
         }
 
         return childElementCount;
+    },
+
+    hasChildElements: node => !!nodeSnapshotPropertyInitializers.childElementCount(node)
+};
+
+export class NodeSnapshot {
+    constructor (node) {
+        this._initializeProperties(node, NODE_SNAPSHOT_PROPERTIES, nodeSnapshotPropertyInitializers);
     }
-}
 
-export class ElementSnapshot extends NodeSnapshot {
-    constructor (element) {
-        super(element);
+    _initializeProperties (node, properties, initializers) {
+        for (var i = 0; i < properties.length; i++) {
+            var property    = properties[i];
+            var initializer = initializers[property];
 
-        this.tagName            = element.tagName.toLowerCase();
-        this.visible            = positionUtils.isElementVisible(element);
-        this.focused            = document.activeElement === element;
-        this.attributes         = ElementSnapshot._getAttrsDictionary(element);
-        this.boundingClientRect = ElementSnapshot._getBoundingClientRect(element);
-        this.classNames         = ElementSnapshot._getClassNames(element);
-        this.style              = ElementSnapshot._getStyle(element);
-        this.innerText          = getInnerText(element);
-
-        [
-            'namespaceURI', 'id',
-            'value', 'checked', 'selected',
-            'scrollWidth', 'scrollHeight', 'scrollLeft', 'scrollTop',
-            'offsetWidth', 'offsetHeight', 'offsetLeft', 'offsetTop',
-            'clientWidth', 'clientHeight', 'clientLeft', 'clientTop'
-        ].forEach(prop => {
-            this[prop] = element[prop];
-        });
+            this[property] = initializer ? initializer(node) : node[property];
+        }
     }
+}
+
 
-    static _getBoundingClientRect (element) {
+// Element
+var elementSnapshotPropertyInitializers = {
+    tagName: element => element.tagName.toLowerCase(),
+    visible: positionUtils.isElementVisible,
+    focused: element => document.activeElement === element,
```

Comment:
it's better to use `domUtils.getActiveElement()` here instead of `document.activeElement` because of [this](https://github.com/DevExpress/testcafe-hammerhead/blob/6e687592ef464e70343e4135596fed38a01881cb/src/client/utils/dom.js#L83)